Sources/BSON/Document.swift

```diff
@@ -1,77 +1,156 @@
 
     private var _buffer: ByteBuffer
 
+    internal var byteLength: Int {
+        guard let byteLength = self._buffer.getInteger(at: 0, endianness: .little, as: Int32.self) else {
+            fatalError("Cannot read byteLength of BSON from buffer")
+        }
+        return Int(byteLength)
+    }
+
     /// An unordered set containing the keys in this document.
-    private var keySet: Set<String>
+    internal private(set) var keySet: Set<String>
 
     internal init(_ elements: [BSON]) { fatalError("Unimplemented") }
 
-    internal init(keyValuePairs: [(String, BSON)]) { fatalError("Unimplemented") }
+    internal init(keyValuePairs: [(String, BSON)]) {
+        self.keySet = Set(keyValuePairs.map { $0.0 })
+        guard self.keySet.count == keyValuePairs.count else {
+            fatalError("Dictionary \(keyValuePairs) contains duplicate keys")
+        }
+
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 100)
+
+        guard !self.keySet.isEmpty else {
+            self = BSONDocument()
+            return
+        }
+
+        let start = self._buffer.writerIndex
+
+        // reserve space for our byteLength that will be calculated
+        self._buffer.writeInteger(0, endianness: .little, as: Int32.self)
+
+        for (key, value) in keyValuePairs {
+            self._buffer.writeInteger(value.bsonValue.bsonType.rawValue, as: UInt8.self)
+            self._buffer.writeCString(key)
+            value.bsonValue.write(to: &self._buffer)
+        }
+        // BSON null terminator
+        self._buffer.writeInteger(0, as: UInt8.self)
+
+        guard let byteLength = Int32(exactly: self._buffer.writerIndex - start) else {
+            fatalError("Data is \(self._buffer.writerIndex - start) bytes, "
+                + "but maximum allowed BSON document size is \(Int32.max) bytes")
+        }
+        // Set byteLength in reserved space
+        self._buffer.setInteger(byteLength, at: 0, endianness: .little, as: Int32.self)
+    }
 
-    /// Initializes a new, empty `Document`.
-    public init() { fatalError("Unimplemented") }
+    /// Initializes a new, empty `BSONDocument`.
+    public init() {
+        self.keySet = Set()
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 5)
+        self._buffer.writeInteger(5, endianness: .little, as: Int32.self)
+        self._buffer.writeBytes([0])
+    }
 
     /**
-     * Initializes a new `Document` from the provided BSON data.
+     * Initializes a new `BSONDocument` from the provided BSON data.
      *
      * - Throws:
      *   - `InvalidArgumentError` if the data passed is invalid BSON
      *
      * - SeeAlso: http://bsonspec.org/
      */
-    public init(fromBSON bson: Data) throws { fatalError("Unimplemented") }
+    public init(fromBSON bson: Data) throws {
+        var buffer = BSON_ALLOCATOR.buffer(capacity: bson.count)
+        buffer.writeBytes(bson)
+        try BSONDocument.validate(buffer)
+        self = BSONDocument(fromUnsafeBSON: buffer)
```

Comment:
these two lines can be replaced with a call to the `fromBSON: ByteBuffer` initializer.